### ISSUE TITLE
Why do Diffusers schedulers produce lower quality outputs compared to ComfyUI?

### ISSUE THREAD

**Issue by nducthang** [huggingface/diffusers#8683]
### Discussed in https://github.com/huggingface/diffusers/discussions/8682

<sup>Originally posted by **nducthang** June 24, 2024</sup>
Hi,

I'm encountering an issue when comparing the quality of ComfyUI and Diffusers. I've noticed that the output of Diffusers is consistently lower than ComfyUI in many cases, despite using the same settings and seed. For the base Diffusers, I've utilized: https://github.com/huggingface/diffusers/blob/main/examples/community/lpw_stable_diffusion_xl.py.

Upon closer inspection, I've identified differences in the scheduler/ksampler between the two base codes. I've also observed variations in CLIP Embedding between the two base codes, but in my experiments, this hasn't significantly impacted the output. The main issue seems to lie with the KSampler.

Has anyone else encountered this issue or have any ideas on improving the Scheduler algorithm of Diffusers?

Here are some prompts I've experimented:
Model: RVXL - Size: (896, 1152)
Positive prompt:
```
female, attractive woman, pretty middle-aged woman, thick hair, (((Caucasian, European, Scandinavian female))), ((hazel eyes, HazelEyed)). (Brunette (Light-Brown-Hair)), ((((long rectangular face, elongated face, oblong face shape, angular chiseled face)), ((wide jaw, big strong chin)))). (((1980s magazine advertisement. Living room. CRT Televesion. 1980s aesthetic. 1980s interior design.))) [object Object] . high quality, dim lighting, soft lighting, sharp focus, f5.6, dslr, High Detail, detailed, ((wide shot))
```
Negative prompt:
```
(((male))), (small chin, receding-chin, puffy face), (((Asian, Chinese, Korean, Japanese, Indian, Pakistani, Black, African, Persian, Arab, Middle Eastern, Hispanic, Latino))), (small chin, receding-chin, puffy face), (blurry), (BadDream:1.2), (UnrealisticDream:1.2), ((bad-hands-5)), (strabismus, cross-eyed:1.2), (signature, watermark, name), (worst quality, poor quality, low quality), ((deformed)), (extra limbs), (extra arms), (extra legs), disfigured, malformed, (nude:1.4), (naked:1.4), (nsfw:1.4), (bikini:1.4), (lingerie:1.4), (underwear:1.4), (teen:1.4), (tween:1.4), (teenage:1.4), (kid:1.6), (child:1.6), (topless, shirtless:1.4), (((greyscale))), (cleavage:1.2), (nipples:1.4)
```

**Comment by sayakpaul** [comment]
Please provide the `diffusers` code you're using so that we can guide you accordingly. Cc: @asomoza

**Comment by asomoza** [comment]
Also some example images with what you consider a good result (automatic1111) and a bad result (diffusers) but please can you change that prompt to not being human beings.

**Comment by nducthang** [comment]
@sayakpaul @asomoza  Thank you for responses.
This is code that I used: 
```
from lpw_stable_diffusion_xl import SDXLLongPromptWeightingPipeline
import torch
from diffusers import DPMSolverMultistepScheduler, EulerAncestralDiscreteScheduler


SD_XL_REALVISXL_V4_0_MODEL_ID = "SG161222/RealVisXL_V4.0"
prompt = "female, attractive woman, pretty middle-aged woman, thick hair, (((Caucasian, European, Scandinavian female))), ((hazel eyes, HazelEyed)). (Brunette (Light-Brown-Hair)), ((((long rectangular face, elongated face, oblong face shape, angular chiseled face)), ((wide jaw, big strong chin)))). (1980s magazine advertisement. Living room. CRT Televesion. 1980s aesthetic. 1980s interior design.) [object Object] . high quality, dim lighting, soft lighting, sharp focus, f5.6, dslr, High Detail, detailed, ((wide shot))"
negative_prompt = "(((male))), (small chin, receding-chin, puffy face), (((Asian, Chinese, Korean, Japanese, Indian, Pakistani, Black, African, Persian, Arab, Middle Eastern, Hispanic, Latino))), (small chin, receding-chin, puffy face), (blurry), (BadDream:1.2), (UnrealisticDream:1.2), ((bad-hands-5)), (strabismus, cross-eyed:1.2), (signature, watermark, name), (worst quality, poor quality, low quality), ((deformed)), (extra limbs), (extra arms), (extra legs), disfigured, malformed, (nude:1.4), (naked:1.4), (nsfw:1.4), (bikini:1.4), (lingerie:1.4), (underwear:1.4), (teen:1.4), (tween:1.4), (teenage:1.4), (kid:1.6), (child:1.6), (topless, shirtless:1.4), (((greyscale))), (cleavage:1.2), (nipples:1.4)"

pipe = SDXLLongPromptWeightingPipeline.from_pretrained(
    SD_XL_REALVISXL_V4_0_MODEL_ID,
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True,
).to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config, use_karras_sigmas=True)
# pipe.schduler = EulerAncestralDiscreteScheduler.from_config(pipe.scheduler.config)
# for seed in [0, 5, 100, 1024]:
for seed in [24483]:
    torch.manual_seed(seed)
    generator = [torch.Generator(device="cuda").manual_seed(seed)]
    image = pipe(
        prompt=prompt,
        negative_prompt=negative_prompt,
        generator=generator,
        num_inference_steps=30,
        denoising_end=1.0,
        guidance_scale=3.5,
        width=832,
        height=1216,
        output_type="pil",
    ).images[0]

    image.save(f"RVXL-{seed}.png")
```
Results of Diffusers and ComfyUI:
![fast_test](https://github.com/huggingface/diffusers/assets/42518675/fff3ad77-49a2-4096-950d-33c49dde4e3f)
![ComfyUI_temp_rsilx_00019_](https://github.com/huggingface/diffusers/assets/42518675/518ec946-fe54-45ef-a6ad-70014b445689)
Somtime, if change seeds, Diffusers base can be generate a TV:
![fast_test-3](https://github.com/huggingface/diffusers/assets/42518675/e570ef45-d0bc-4184-8802-3ebdb27f97ff)

With ComfyUI I see stable results and this phenomenon does not occur.

**Comment by nducthang** [comment]
Sometimes images get corrupted (random seed) in diffusers base. I tried change scheduler but the result are not improved
![fast_test-33](https://github.com/huggingface/diffusers/assets/42518675/61bec9cd-5be0-4c83-ab85-70201bdd6f27)

**Comment by asomoza** [comment]
is there a reason why you're using such a low cfg?

**Comment by nducthang** [comment]
> is there a reason why you're using such a low cfg?

When I tested before and decided to use DPM 2a Karras I tested a cfg of 3.5 and 7.  The 3.5 seemed to give better results than 7. 7 started causing there to be artifacts even at 40 steps.  30 steps with a cfg of 3.5 seemed to be the sweet spot.

**Comment by asomoza** [comment]
ok, I see that you're getting some weird results also, I'll do some tests but this doesn't seem to be an issue with the schedulers, I can generate just fine with your prompt (truncated) and without artifacts. I'll test with the long weighting prompts.

**Comment by nducthang** [comment]
> ok, I see that you're getting some weird results also, I'll do some tests but this doesn't seem to be an issue with the schedulers, I can generate just fine with your prompt (truncated) and without artifacts. I'll test with the long weighting prompts.
> 
> With `DPMSolverMultistepScheduler` I always use a cfg of 6.5 and 25 steps and works fine, so its weird you're getting artifacts with more steps.
> 
> This reminds me an issue we had with the schedulers way back, are you using the latest release?

Yes, I also find that the diffusers can be stable with shorter prompt and without weighting. However, If I evaluate emotionally, I think the quality doesn't seem to be as good as ComfyUI. So, I think shedulers might be the cause.

In my experiment, I am using lastest release.

**Comment by asomoza** [comment]
I'm definitely not getting your bad results, this is the kind I get:

|||||
|---|---|---|---|
|![20240625001057_2154947393](https://github.com/huggingface/diffusers/assets/5442875/eb459f6c-5675-45b1-a9e4-dcb59a65bb91)|![20240625001203_3360672778](https://github.com/huggingface/diffusers/assets/5442875/3b147af5-06da-4c51-a9f0-bfb8dec2696a)|![20240625001104_2328714939](https://github.com/huggingface/diffusers/assets/5442875/eaf10319-b1bb-4eb3-a666-e674bc0ae41f)|![20240625001210_3362655055](https://github.com/huggingface/diffusers/assets/5442875/14295ded-14cf-4c7b-b005-86e486751c07)|

I did lots and never got the artifacts or the weird television thing.

**Comment by asomoza** [comment]
Also I used the new library by the same author @xhinker as the pipeline:

https://github.com/xhinker/sd_embed

But the results should be the same I think, if not, maybe you can try using that.

**Comment by asomoza** [comment]
Also do I need to use something in comfyui to be able to use those prompts? If I just use the vanilla nodes I get worse results than diffusers in my opinion, they look kinda fake and plastic with the exact same params:

|comfyui|diffusers|
|---|---|
|![ComfyUI_00271_](https://github.com/huggingface/diffusers/assets/5442875/0dfa0f98-47db-4794-9034-86052484ae8c)|![20240625002932_237369457047662](https://github.com/huggingface/diffusers/assets/5442875/a1c3595a-1605-4839-8d6e-8cce437d093e)|

**Comment by nducthang** [comment]
@asomoza  Do you used SDXLLongPromptWeightingPipeline form https://github.com/huggingface/diffusers/blob/main/examples/community/lpw_stable_diffusion_xl.py ?

Seed 0, 100 and 1024 (CFG: 3.5) (Same results with CFG 6.5):
![RVXL-0](https://github.com/huggingface/diffusers/assets/42518675/34a50e16-971a-4d28-bcc2-c2dfe3236773)
![RVXL-100](https://github.com/huggingface/diffusers/assets/42518675/c4140b51-7efe-4160-8362-5c2553db97cb)
![RVXL-1024](https://github.com/huggingface/diffusers/assets/42518675/5abd5920-9444-45c7-8153-997b2537b5cb)

If use choose width=1024 and height=1536. The results will be worse.

![RVXL-0-1](https://github.com/huggingface/diffusers/assets/42518675/782579b9-bea0-4aea-ad56-010d85653328)
![RVXL-100-1](https://github.com/huggingface/diffusers/assets/42518675/9adfcf02-e7ce-4967-883f-4e3b5b464b45)
![RVXL-1024-1](https://github.com/huggingface/diffusers/assets/42518675/12e3059a-a430-4a35-9a0b-4ed72979a8d4)

**Comment by asomoza** [comment]
> Do you used SDXLLongPromptWeightingPipeline

No, I used the [library](https://github.com/xhinker/sd_embed) I posted which is from the same author which enables the use of long weighted prompts without the need to load any specific pipeline.

Also take into account that you can't just use any resolution, to get the best results you'll need to use the resolution the model was trained before, and mostly they recommend that they match the number of pixels of 1024x1024, so if you go high with one it's better to lower the other one. 

1024x1536 goes way higher than that, I can get good results with it though but I also start getting bad ones, but the same happens with any UI, is something related to the model.

**Comment by nducthang** [comment]
I attached my workflow in ComfyUI. If look kinda fake and plastic, you can use CFG 3.5.
[workflow (2).json](https://github.com/user-attachments/files/15964795/workflow.2.json)

The strange thing is that I find the pipeline diffusers I use give unstable results. I will try the library https://github.com/xhinker/sd_embed that you suggested to see if the quality improves, thank you!

**Comment by xhinker** [comment]
I tried your prompt. 
Using 
```py
import gc
import torch
from diffusers import StableDiffusionXLPipeline
from sd_embed.embedding_funcs import get_weighted_text_embeddings_sdxl
from diffusers.utils import make_image_grid

model_path = "SG161222/RealVisXL_V4.0"
pipe = StableDiffusionXLPipeline.from_pretrained(
    model_path,
    torch_dtype=torch.float16
)

#%%
pipe.to('cuda')
prompt = """\
female, attractive woman, pretty middle-aged woman, thick hair, (((Caucasian, European, Scandinavian female))), ((hazel eyes, HazelEyed)). (Brunette (Light-Brown-Hair)), ((((long rectangular face, elongated face, oblong face shape, angular chiseled face)), ((wide jaw, big strong chin)))). (((1980s magazine advertisement. Living room. CRT Televesion. 1980s aesthetic. 1980s interior design.))) [object Object] . high quality, dim lighting, soft lighting, sharp focus, f5.6, dslr, High Detail, detailed, ((wide shot))
"""

neg_prompt = """\
(((male))), (small chin, receding-chin, puffy face), (((Asian, Chinese, Korean, Japanese, Indian, Pakistani, Black, African, Persian, Arab, Middle Eastern, Hispanic, Latino))), (small chin, receding-chin, puffy face), (blurry), (BadDream:1.2), (UnrealisticDream:1.2), ((bad-hands-5)), (strabismus, cross-eyed:1.2), (signature, watermark, name), (worst quality, poor quality, low quality), ((deformed)), (extra limbs), (extra arms), (extra legs), disfigured, malformed, (nude:1.4), (naked:1.4), (nsfw:1.4), (bikini:1.4), (lingerie:1.4), (underwear:1.4), (teen:1.4), (tween:1.4), (teenage:1.4), (kid:1.6), (child:1.6), (topless, shirtless:1.4), (((greyscale))), (cleavage:1.2), (nipples:1.4)
"""

(
    prompt_embeds
    , prompt_neg_embeds
    , pooled_prompt_embeds
    , negative_pooled_prompt_embeds
) = get_weighted_text_embeddings_sdxl(
    pipe
    , prompt = prompt
    , neg_prompt = neg_prompt
)

image = pipe(
    prompt_embeds                   = prompt_embeds
    , negative_prompt_embeds        = prompt_neg_embeds
    , pooled_prompt_embeds          = pooled_prompt_embeds
    , negative_pooled_prompt_embeds = negative_pooled_prompt_embeds
    , num_inference_steps           = 30
    , height                        = 1024 
    , width                         = 1024
    , guidance_scale                = 7.0
    , generator                     = torch.Generator("cuda").manual_seed(2)
    , num_images_per_prompt         = 4
).images
# display(image)

del prompt_embeds, prompt_neg_embeds,pooled_prompt_embeds, negative_pooled_prompt_embeds
pipe.to('cpu')
gc.collect()
torch.cuda.empty_cache()

make_image_grid(image,rows=2, cols=2)
```
I got:
![image](https://github.com/huggingface/diffusers/assets/7898116/24d1afbf-ec30-47ea-930b-061bac6111f8)

**Comment by xhinker** [comment]
@nducthang for `lpw_stable_diffusion_xl` the community added some additional logic to support more features. 
I found that if you remove the `[object, object]` from your prompt, you will get better result like this: 
```py
from diffusers import DiffusionPipeline
import torch
from diffusers.utils import make_image_grid

model_path = "SG161222/RealVisXL_V4.0"

pipe = DiffusionPipeline.from_pretrained(
    model_path
    , torch_dtype       = torch.float16
    , use_safetensors   = True
    , variant           = "fp16"
    , custom_pipeline   = "lpw_stable_diffusion_xl",
)

#%%
pipe.to("cuda")
prompt = """\
female, attractive woman, pretty middle-aged woman, thick hair, 
(((Caucasian, European, Scandinavian female))), ((hazel eyes, HazelEyed)). 
(Brunette (Light-Brown-Hair)), ((((long rectangular face, elongated face, oblong face shape, 
angular chiseled face)), ((wide jaw, big strong chin)))). (((1980s magazine advertisement. 
Living room. CRT Televesion. 1980s aesthetic. 1980s interior design.))). 
high quality, dim lighting, soft lighting, sharp focus, f5.6, dslr, High Detail, detailed, 
((wide shot))
"""

neg_prompt = """\
(((male))), (small chin, receding-chin, puffy face), 
(((Asian, Chinese, Korean, Japanese, Indian, Pakistani, Black, African, Persian, Arab, Middle Eastern, Hispanic, Latino))), 
(small chin, receding-chin, puffy face), (blurry), (BadDream:1.2), 
(UnrealisticDream:1.2), ((bad-hands-5)), (strabismus, cross-eyed:1.2), 
(signature, watermark, name), (worst quality, poor quality, low quality),
((deformed)), (extra limbs), (extra arms), (extra legs), disfigured, malformed, 
(nude:1.4), (naked:1.4), (nsfw:1.4), (bikini:1.4), (lingerie:1.4), (underwear:1.4), 
(teen:1.4), (tween:1.4), (teenage:1.4), (kid:1.6), (child:1.6), 
(topless, shirtless:1.4), (((greyscale))), (cleavage:1.2), (nipples:1.4)
"""

images = pipe(
    prompt                  = prompt 
    , negative_prompt       = neg_prompt 
    , generator             = torch.Generator('cuda').manual_seed(3)
    , height                = 1024
    , width                 = 1024 
    , num_inference_steps   = 30
    , guidance_scale        = 7.0
    , num_images_per_prompt = 4
).images

pipe.to("cpu")
torch.cuda.empty_cache()
# images

make_image_grid(images,rows=2, cols=2)
```
Result:
![image](https://github.com/huggingface/diffusers/assets/7898116/ba168ed3-e720-484a-8a05-b04766e7a939)

**Comment by xhinker** [comment]
In your prompt, I saw 
```
...
(((1980s magazine advertisement. 
Living room. CRT Televesion. 1980s aesthetic. 1980s interior design.)))
...
```
I think Diffusers generates results more aligned than the comfy UI version 😊

**Comment by asomoza** [comment]
yeah, what was weird is that it was floating in the middle 😊

Funny enough, I don't get the television that often though, also with your post I found that it's misspelled. If I fix it I start to get the television.

**Comment by asomoza** [comment]
Also I can't get good results with comfyui, even with low CFG, they all look plastic and fake and also, now, I can clearly see that it doesn't respect half of the prompt. Probably comfyui is not interpreting the prompt correctly, also maybe it's just doing an average with the long prompts which makes the generation concepts more diluted.

**Comment by nducthang** [comment]
I got better results using `sd_embed` . I also found some problems in my prompt, and also see that  Diffusers generates results more aligned than the comfyUI. Thanks @asomoza @xhinker and @sayakpaul
